File: Java/LimitsReporting/src/com/matthey/pmm/limits/reporting/translated/LiquidityLimitsChecker.java
package com.matthey.pmm.limits.reporting.translated;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.openlink.util.logging.PluginLog;

public class LiquidityLimitsChecker {
    private static final String runType = "Liquidity";

    private final LimitsReportingConnector connector;
    
    public LiquidityLimitsChecker (final LimitsReportingConnector connector) {
    	this.connector = connector;
    }
    
    public List<RunResult> check() {
        PluginLog.info("checking liquidity limits");
        List<RunResult> runResults = new ArrayList<RunResult>();
        
        for (LiquidityLimit liquidityLimit : connector.getLiquidityLimits()) {
            PluginLog.info("liquidity limit:" + liquidityLimit);
            double fxRate = connector.getGbpUsdRate(connector.getRunDate());
            PluginLog.info("GBP/USD rate: " + fxRate);
            Map<String, Double> metalPrices = connector.getMetalPrices();
            PluginLog.info("metal prices: " + metalPrices);
            MetalBalances metalBalances = connector.getMetalBalances();
            int liquidity = metalBalances.getBalance(runType, liquidityLimit.getMetal());
            boolean breach = !(liquidity >= liquidityLimit.getLowerLimit() &&
            				   liquidity <= liquidityLimit.getUpperLimit());
            boolean breachLowerLimit = liquidity < liquidityLimit.getLowerLimit();
            int diff = !breach?0:(breachLowerLimit?
            		liquidityLimit.getLowerLimit() - liquidity:
            		liquidity - liquidityLimit.getUpperLimit());
            double breachTOz = getBreachBalancesInTOz(metalBalances, liquidityLimit.getMetal(), diff);
            PluginLog.info("breach in TOz: " + breachTOz);
            double liability = breachTOz * metalPrices.get(liquidityLimit.getMetal()) / fxRate;
            PluginLog.info("liability: " + liability);
            boolean critical = !breachLowerLimit && (liability > liquidityLimit.getMaxLiability());
            String liquidityBreachLimit = breach?(breachLowerLimit?"Lower Limit":"Upper Limit"):"";
            
            RunResult runResult =   new RunResult (
            		connector.getRunDate(), // runTime
            		runType, // runType
            		"", // desk
            		liquidityLimit.getMetal(), // metal
            		liquidityLimit.getLowerLimit(), // liquidityLowerLimit
            		liquidityLimit.getUpperLimit(), // liquidityUpperLimit
            		liquidityLimit.getMaxLiability(), // liquidityMaxLiability
            		0, // position limit
            		breach, // breach
            		liquidityBreachLimit, // liquidityBreachLimit
            		liquidity, // currentPosition
            		diff, // liquidityDiff
            		breachTOz, // breachTOz
            		liability, // breachGBP
            		critical,  // critical
            		RunResult.getBreachDates(connector, breach, runType, 
            				liquidityBreachLimit, "", liquidityLimit.getMetal()) // breachDates
            		);
            runResults.add(runResult);
        }
        return runResults;
    }

    private double getBreachBalancesInTOz(MetalBalances metalBalances, String metal, int diff) {
    	List<String> balancesLinesTitlesForRunType = new ArrayList<>();
    	for (BalanceLine balanceLine : connector.getBalanceLines()) {
    		if (balanceLine.getPurpose().equals(runType)) {
    			balancesLinesTitlesForRunType.add(balanceLine.getLineTitle());
    		}
    	}    	
        PluginLog.info("balance lines to be used " + balancesLinesTitlesForRunType);
        int totalBalances = 0;
        for (String balanceLineTitle : balancesLinesTitlesForRunType) {
            int balance = metalBalances.getBalance(balanceLineTitle, metal);
            totalBalances += balance;
        }
        PluginLog.info("total balance for " + metal + " in TOz:" + totalBalances);
        return totalBalances * diff / 100.0d;
    }
}

File: Java/LimitsReporting/src/main/kotlin/com/matthey/pmm/limits/reporting/LiquidityLimitsChecker.kt
package com.matthey.pmm.limits.reporting

import org.slf4j.LoggerFactory

class LiquidityLimitsChecker(private val connector: LimitsReportingConnector) {

    private val logger = LoggerFactory.getLogger(javaClass)

    private val runType = "Liquidity"

    fun check(): List<RunResult> {
        logger.info("checking liquidity limits")

        return connector.liquidityLimits.map {
            logger.info("liquidity limits: $it")
            val fxRate = connector.getGbpUsdRate()
            logger.info("GBP/USD rate: $fxRate")
            val metalPrices = connector.metalPrices
            logger.info("metal prices: $metalPrices")

            val metalBalances = connector.metalBalances
            val liquidity = metalBalances.getBalance(runType, it.metal)
            val breach = liquidity !in it.lowerLimit..it.upperLimit
            val breachLowerLimit = liquidity < it.lowerLimit
            val diff =
                if (!breach) 0 else if (breachLowerLimit) it.lowerLimit - liquidity else liquidity - it.upperLimit
            val breachTOz = getBreachBalancesInTOz(metalBalances, it.metal, diff)
            logger.info("breach in TOz: $breachTOz")
            val liability = breachTOz * metalPrices.getValue(it.metal) / fxRate
            logger.info("liability: $liability")
            val critical = !breachLowerLimit && liability > it.maxLiability
            val liquidityBreachLimit = if (breach) if (breachLowerLimit) "Lower Limit" else "Upper Limit" else ""

            RunResult(
                runTime = connector.runDate,
                runType = runType,
                metal = it.metal,
                liquidityLowerLimit = it.lowerLimit,
                liquidityUpperLimit = it.upperLimit,
                liquidityMaxLiability = it.maxLiability,
                breach = breach,
                liquidityBreachLimit = liquidityBreachLimit,
                currentPosition = liquidity.toDouble(),
                liquidityDiff = diff,
                breachTOz = breachTOz,
                breachGBP = liability,
                critical = critical,
                breachDates = RunResult.getBreachDates(
                    connector,
                    breach,
                    runType = runType,
                    liquidityBreachLimit = liquidityBreachLimit,
                    metal = it.metal
                )
            )
        }
    }

    private fun getBreachBalancesInTOz(metalBalances: MetalBalances, metal: String, diff: Int): Double {
        val balanceLines = connector.balanceLines.filter { it.purpose == runType }.map { it.lineTitle }
        logger.info("balance lines to be used $balanceLines")
        val totalBalances = balanceLines.sumBy { metalBalances.getBalance(it, metal) }
        logger.info("total balance for $metal in TOz: $totalBalances")
        return totalBalances * diff / 100.0
    }
}